File: Gearmaker/GearView.m
//
//  GearView.m
//  Gearmaker
//
//  Created by Torsten Kammer on 28.01.13.
//  Copyright (c) 2013 Torsten Kammer. All rights reserved.
//

#import "GearView.h"

#import "Gear.h"

@interface GearView ()

@property (nonatomic) NSArray *outlinePoints;

@end

@implementation GearView

- (id)initWithFrame:(NSRect)frame
{
	if (!(self = [super initWithFrame:frame])) return nil;
    
	[self addObserver:self forKeyPath:@"gear.module" options:NSKeyValueObservingOptionNew context:0];
	[self addObserver:self forKeyPath:@"gear.teeth" options:NSKeyValueObservingOptionNew context:0];
	[self addObserver:self forKeyPath:@"gear.eingriffwinkel" options:NSKeyValueObservingOptionNew context:0];
	[self addObserver:self forKeyPath:@"gear.kopfspielfaktor" options:NSKeyValueObservingOptionNew context:0];
	
    return self;
}

- (void)dealloc
{
	[self removeObserver:self forKeyPath:@"gear.module"];
	[self removeObserver:self forKeyPath:@"gear.teeth"];
	[self removeObserver:self forKeyPath:@"gear.eingriffwinkel"];
	[self removeObserver:self forKeyPath:@"gear.kopfspielfaktor"];
}

- (void)observeValueForKeyPath:(NSString *)keyPath ofObject:(id)object change:(NSDictionary *)change context:(void *)context
{
	self.outlinePoints = [self.gear getOutlinePointsAtInveral:1.0];
	self.needsDisplay = YES;
}

- (void)drawRect:(NSRect)dirtyRect
{
	// Clear
	[[NSColor whiteColor] set];
	NSBezierPath *rectPath = [NSBezierPath bezierPathWithRect:dirtyRect];
	[rectPath fill];
	
	if (self.outlinePoints.count == 0) return;
	
	// Prepare translation
	NSAffineTransform *transform = [NSAffineTransform transform];
	[transform translateXBy:NSMidX(self.bounds) yBy:NSMidY(self.bounds)];
	
	// Draw teilkreis
	[[NSColor redColor] set];
	CGFloat teilkreisRadius = self.gear.teilkreisdurchmesser * 0.5;
	NSBezierPath *teilkreisPath = [NSBezierPath bezierPathWithOvalInRect:NSMakeRect(-teilkreisRadius, -teilkreisRadius, teilkreisRadius*2., teilkreisRadius*2.)];
	[[transform transformBezierPath:teilkreisPath] stroke];
	
	// Draw gear
	[[NSColor blackColor] set];
	NSBezierPath *gearPath = [NSBezierPath bezierPath];
	[gearPath moveToPoint:[self.outlinePoints[0] pointValue]];
	for (NSValue *point in self.outlinePoints)
		[gearPath lineToPoint:point.pointValue];
	[gearPath closePath];
	[[transform transformBezierPath:gearPath] stroke];
}

@end

File: Gearmaker/GearView.swift
//
//  GearView.swift
//  Gearmaker
//
//  Created by Torsten Kammer on 13.11.22.
//  Copyright © 2022 Torsten Kammer. All rights reserved.
//

import Cocoa

class GearView: NSView {
    
    @objc var gear: Gear? = nil
    private var observations: [NSKeyValueObservation] = []
    private var outlinePoints: [SIMD2<Double>] = []
    
    override init(frame frameRect: NSRect) {
        super.init(frame: frameRect)
        
        prepareObservations()
    }
    
    required init?(coder: NSCoder) {
        super.init(coder: coder)
        
        prepareObservations()
    }
    
    private func prepareObservations() {
        observations.append(self.observe(\.gear?.module, changeHandler: { view, _ in
            view.updateGear()
        }))
        observations.append(self.observe(\.gear?.teeth, changeHandler: { view, _ in
            view.updateGear()
        }))
        observations.append(self.observe(\.gear?.eingriffwinkel, changeHandler: { view, _ in
            view.updateGear()
        }))
        observations.append(self.observe(\.gear?.kopfspielfaktor, changeHandler: { view, _ in
            view.updateGear()
        }))
    }
    
    override func draw(_ dirtyRect: NSRect) {
        // Clear background
        NSColor.white.set()
        NSBezierPath(rect: dirtyRect).fill()
        
        guard let gear = gear, !outlinePoints.isEmpty else {
            return
        }
        
        // Translate to draw from center
        let transform = NSAffineTransform()
        transform.translateX(by: bounds.midX, yBy: bounds.midY)
        
        // Draw teilkreis
        NSColor.red.set()
        let teilkreisradius = gear.teilkreisdurchmesser / 2
        let teilkreispath = NSBezierPath(ovalIn: NSRect(origin: CGPoint(x: -teilkreisradius, y: -teilkreisradius), size: CGSize(width: gear.teilkreisdurchmesser, height: gear.teilkreisdurchmesser)))
        transform.transform(teilkreispath).stroke()
        
        // Draw gear
        NSColor.black.set()
        let gearPath = NSBezierPath()
        gearPath.move(to: CGPoint(point: outlinePoints.first!))
        for point in outlinePoints {
            gearPath.line(to: CGPoint(point: point))
        }
        gearPath.close()
        transform.transform(gearPath).stroke()
    }
                         
    private func updateGear() {
        outlinePoints = gear?.outlinePoints(interval: 1.0) ?? []
        needsDisplay = true
    }
}
